contracts/EpochTokenLocker.sol

```diff
@@ -0,0 +1,151 @@
+// which becomes claimable after the current epoch expired.
+
+
+contract EpochTokenLocker {
+    using SafeMath for uint;
+
+
+    event Deposit(
+        address user,
+        address token,
+        uint amount,
+        uint stateIndex
+    );
+
+    event WithdrawRequest(
+        address user,
+        address token,
+        uint amount,
+        uint stateIndex
+    );
+
+    event Withdraw(
+        address user,
+        address token,
+        uint amount
+    );
+
+    // User => Token => BalanceState
+    mapping(address => mapping(address => BalanceState)) private balanceStates;
+
+    struct BalanceState {
+        uint256 balance;
+        PendingFlux pendingDeposits; // deposits will be credited in any next epoch, i.e. currentStateIndex > stateIndex
+        PendingFlux pendingWithdraws; // withdraws are allowed in any next epoch , i.e. currentStateIndex > stateIndex
+    }
+
+    struct PendingFlux {
+        uint256 amount;
+        uint256 stateIndex;
+    }
+
+    uint256 public currentStateIndex = 0;
+
+    function updateBalance(address user, address token) public returns(uint256) {
+        updateDepositsBalance(user, token);
+        uint balance = balanceStates[user][token].balance;
+        if (balanceStates[user][token].pendingWithdraws.stateIndex < currentStateIndex) {
+            balance -= Math.min(balanceStates[user][token].pendingWithdraws.amount, balance);
+        }
+        return balance;
+    }
+
+    function deposit(address token, uint amount) public {
+        updateDepositsBalance(msg.sender, token);
+        require(
+            ERC20(token).transferFrom(msg.sender, address(this), amount),
+            "Tokentransfer for deposit was not successful"
+        );
+        balanceStates[msg.sender][token].pendingDeposits.amount = balanceStates[msg.sender][token].pendingDeposits.amount
+            .add(amount);
+        balanceStates[msg.sender][token].pendingDeposits.stateIndex = currentStateIndex;
+        emit Deposit(msg.sender, token, amount, currentStateIndex);
+    }
+
+    function requestWithdraw(address token, uint amount) public {
+        balanceStates[msg.sender][token].pendingWithdraws = PendingFlux({ amount: amount, stateIndex: currentStateIndex });
+        emit WithdrawRequest(msg.sender, token, amount, currentStateIndex);
+    }
+
+    function withdraw(address token) public {
+        updateDepositsBalance(msg.sender, token); // withdrawn amount might just be deposited before
+
+        require(
+            balanceStates[msg.sender][token].pendingWithdraws.stateIndex < currentStateIndex,
+            "withdraw was not registered previously"
+        );
+
+        uint amount = Math.min(
+            balanceStates[msg.sender][token].balance,
+            balanceStates[msg.sender][token].pendingWithdraws.amount
+        );
+
+        balanceStates[msg.sender][token].balance = balanceStates[msg.sender][token].balance.sub(amount);
```

Comment:
Lines `92-96` I think is best if they are executed only if `amount > 0`.
I guess is not super important, but we avoid emitting events of `amount = 0`

Reply:
Yeah, for me it is hard to foresee exactly how the events are used and so it is hard to code it properly.
If we want to be able to calculate everything from events, we might want to have the event even for amounts = 0. Because then, we need to know from somewhere, that the pendingWithdraws was deleted in order to calculate the balance correctly in the future. And currently the only way to know it is from the event.